### PR TITLE
applications: asset_tracker_v2: Add support for QEMU x86

### PR DESCRIPTION
**Pull request by simensrostad**
Add support for QEMU x86

### CODE REVIEW

File: applications/asset_tracker_v2/overlay-azure.conf [hunk 1]
```diff
@@ -10,14 +10,11 @@ CONFIG_AZURE_IOT_HUB_AUTO_DEVICE_TWIN_REQUEST=y
 CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP=y
 CONFIG_AZURE_IOT_HUB_MQTT_RX_TX_BUFFER_LEN=2048
 CONFIG_AZURE_IOT_HUB_STACK_SIZE=4096
-CONFIG_AZURE_FOTA=y
-CONFIG_AZURE_FOTA_TLS=y
```
Comment (coderbyheart) [inline applications/asset_tracker_v2/overlay-azure.conf:14]: **question:** why is this removed?

Comment (coderbyheart) [reply applications/asset_tracker_v2/overlay-azure.conf:14]: Same for sec tag below.

File: applications/asset_tracker_v2/prj_qemu_x86.conf [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+CONFIG_MBEDTLS_KEY_EXCHANGE_RSA_ENABLED=y
+CONFIG_MBEDTLS_MAC_SHA256_ENABLED=y
+CONFIG_MBEDTLS_ECP_DP_SECP256R1_ENABLED=y
+CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED=y
+
+# DNS
+CONFIG_DNS_RESOLVER=y
+CONFIG_DNS_SERVER_IP_ADDRESSES=y
+## Use Google as default DNS server
+CONFIG_DNS_SERVER1="8.8.8.8"
```
Comment (coderbyheart) [inline applications/asset_tracker_v2/prj_qemu_x86.conf:88]: **suggestion:** use CloudFlare DNS [1.1.1.1](https://1.1.1.1/), which is more privacy friendly.

File: applications/asset_tracker_v2/src/events/debug_module_event.h [hunk 3]
```diff
@@ -34,6 +34,17 @@ enum debug_module_event_type {
...
 	DEBUG_EVT_MEMFAULT_DATA_READY,
 
+	/** Event sent after boot when building for QEMU x86. This event acts as a placeholder for
+	 *  MODEM_EVT_INITIALIZED which is not sent due to the modem module being disabled for
+	 *  QEMU x86 builds.
+	 */
+	DEBUG_EVT_QEMU_X86_INITIALIZED,
+
+	/** Event sent when the application is built for QEMU x86.
+	 *  When built for QEMU x86 it is assumed that the application is connected to network.
```
Comment (coderbyheart) [inline applications/asset_tracker_v2/src/events/debug_module_event.h:44]:
```suggestion
	 *  When built for QEMU x86 it is assumed that the application is connected to the network.
```

File: applications/asset_tracker_v2/src/main.c [hunk 4]
```diff
@@ -33,7 +34,7 @@
 #include <logging/log.h>
 #include <logging/log_ctrl.h>
 
-LOG_MODULE_REGISTER(MODULE, CONFIG_APPLICATION_MODULE_LOG_LEVEL);
+LOG_MODULE_REGISTER(app_test_placeholder, CONFIG_APPLICATION_MODULE_LOG_LEVEL);
```
Comment (coderbyheart) [inline applications/asset_tracker_v2/src/main.c:37]: I have no idea about testing, but I am surprised to see a test placeholder being used in the main application file.

File: applications/asset_tracker_v2/src/main.c [hunk 5]
```diff
@@ -546,7 +549,7 @@ void main(void)
 		k_sleep(K_SECONDS(5));
 		sys_reboot(SYS_REBOOT_COLD);
 	} else {
-		module_set_state(MODULE_STATE_READY);
+		// module_set_state(MODULE_STATE_READY);
```
Comment (coderbyheart) [inline applications/asset_tracker_v2/src/main.c:552]:
```suggestion
```
Remove disabled line.

File: subsys/net/lib/aws_fota/Kconfig [hunk 6]
```diff
@@ -8,6 +8,7 @@ menuconfig AWS_FOTA
 	select AWS_JOBS
 	depends on FOTA_DOWNLOAD
 	depends on CJSON_LIB
+	default y if AWS_IOT
```
Comment (coderbyheart) [inline subsys/net/lib/aws_fota/Kconfig:11]: **nitpick:** this changes the behaviour of the configuration and is not related to QEMU.

**Review by coderbyheart — APPROVED** [review]

File: applications/asset_tracker_v2/src/main.c (hunk 4 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/main.c:37]: The changes to `main.c` is just a workaround needed in order to properly run the app in qemu. The root cause will be fixed with https://github.com/nrfconnect/sdk-nrf/pull/7147.

**Comment by simensrostad** [comment]
Adding DNM label until https://github.com/nrfconnect/sdk-nrf/pull/7147 is merged.

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module [hunk 7]
```diff
@@ -32,12 +32,28 @@ choice NRF_CLOUD_SEND_SOCKET
...
-# MQTT Transport library
-config MQTT_CLEAN_SESSION
-	default y
+choice NRF_CLOUD_CLIENT_ID_SRC
+	default NRF_CLOUD_CLIENT_ID_SRC_RUNTIME
+endchoice
+
+if BOARD_QEMU_X86
+
+config NRF_CLOUD_FOTA
```
Comment (rlubos) [inline applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:41]: Why is it needed to redefine this symbol at application level?

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module [hunk 8]
```diff
@@ -32,12 +32,28 @@ choice NRF_CLOUD_SEND_SOCKET
...
+	bool ""
+	default n
+
+endif # BOARD_QEMU_X86
 
 endif # NRF_CLOUD_MQTT
 
+if AZURE_FOTA_TLS
+
+config AZURE_FOTA_SEC_TAG
```
Comment (rlubos) [inline applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:51]: Ditto.

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_mqtt [hunk 9]
```diff
@@ -68,6 +69,7 @@ config NRF_CLOUD_SEND_NONBLOCKING
 
 config NRF_CLOUD_SEND_TIMEOUT
 	bool "Send data with socket timeout"
+	depends on !BOARD_QEMU_X86
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_mqtt:72]: Why? `SO_SNDTIMEO` is supported by the native stack. There's a Kconfig knob to enable it though (`CONFIG_NET_CONTEXT_SNDTIMEO`).

**Comment by rlubos** [comment]
The dependency has been merged

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_mqtt (hunk 9 above)
Comment (simensrostad) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_mqtt:72]: Wasn't aware of this, will remove the change.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Main observations:
- Help text missing in Kconfig
- CMake include directories must be fixed

File: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c [hunk 10]
```diff
@@ -17,9 +17,9 @@
 #include <net/azure_fota.h>
 #endif
 
-#if defined(CONFIG_BOARD_QEMU_X86)
-#include "certificates.h"
-#endif
+#if defined(CONFIG_AZURE_IOT_HUB_PROVISION_CERTIFICATES)
+#include CONFIG_AZURE_IOT_HUB_CERTIFICATES_FILE
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:21]: I can change it to include `certificates.h`. I agree, its more clear. The change was inspired by how it is done in the nRF Cloud library and was partly done for alignment reasons.

File: subsys/net/lib/azure_iot_hub/Kconfig [hunk 11]
```diff
@@ -38,6 +38,19 @@ config AZURE_IOT_HUB_SEND_TIMEOUT_SEC
...
+	depends on BOARD_QEMU_X86
+	help
+	  Enable run-time provisioning of certificates from the
+	  certificates header file selected by using AZURE_IOT_HUB_CERTIFICATES_FILE
+	  This option is only available for QEMU x86 build targets.
+
+config AZURE_IOT_HUB_CERTIFICATES_FILE
+	string "Azure IoT Hub certificates"
+	depends on AZURE_IOT_HUB_PROVISION_CERTIFICATES
+	default "certificates.h"
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply subsys/net/lib/azure_iot_hub/Kconfig:52]: Ill see if I can cough up some comments in the source code explaining the format of the content of the certificates file.

File: applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c [hunk 12]
```diff
@@ -234,6 +244,27 @@ int cloud_wrap_init(cloud_wrap_evt_handler_t event_handler)
 		.event_handler = nrf_cloud_event_handler,
 	};
 
+#if !defined(CONFIG_CLOUD_CLIENT_ID_USE_CUSTOM)
+	char imei_buf[20 + sizeof("OK\r\n")];
```
Comment (dilner) [inline applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c:248]: Shouldn't the size of imeibuf depend on NRF_CLOUD_ID_LEN?

File: applications/asset_tracker_v2/src/cloud/CMakeLists.txt [hunk 13]
```diff
@@ -4,7 +4,7 @@
 # SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 #
 
-target_include_directories(app PRIVATE .)
+zephyr_include_directories(app PRIVATE .)
```
Comment (tejlmand) [inline applications/asset_tracker_v2/src/cloud/CMakeLists.txt:7]: Seems an earlier comment here somehow was lost.

Why is this suddenly targeting `zephyr_interface` library (and thus a global include folder), and not `app` lib ?

Secondly, the use of `app PRIVATE` is wrong when using `zephyr_include_directories()` or `zephyr_library_include_directories()`.

**Review by tejlmand — CHANGES_REQUESTED** [review]
CMake comment somehow was lost in previous review.

File: subsys/net/lib/aws_fota/Kconfig (hunk 6 above)
Comment (simensrostad) [reply subsys/net/lib/aws_fota/Kconfig:12]: If QEMU is enabled, we want to prevent users of enabling FOTA. Otherwise, FOTA will be enabled as a default feature as long as the FOTA_DOWNLOAD library is enabled. Same for Azure.

File: applications/asset_tracker_v2/overlay-azure.conf (hunk 1 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/overlay-azure.conf:14]: These are no longer necessary. Azure FOTA is auto-enabled as long as the fw is build for a 91 compatible board and the fota download library is enabled.

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module (hunk 7 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.cloud_module]: You're right. This change is not necessary. I updated the `prj_qemu_x86.conf` to explicitly disable `NRF_FOTA`.

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module (hunk 8 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.cloud_module]: I moved this to `Kconfig.azure_fota_patch`. Setting this Kconfig in the `overlay-azure.conf` (where it previously was set) file doesn't work because the Kconfig is not visible for QEMU x86 builds when Azure FOTA is disabled. It has to be set in the cases Azure FOTA is enabled to prevent the build assert in https://github.com/simensrostad/fw-nrfconnect-nrf/blob/9c79728d9daa63264764dd844c252693334684a7/subsys/net/lib/azure_fota/azure_fota.c#L82.

File: subsys/net/lib/azure_iot_hub/Kconfig (hunk 11 above)
[the remark this answers is not in the document]
Comment (simensrostad) [reply subsys/net/lib/azure_iot_hub/Kconfig]: Run-time provisioning of certificates are now aligned between all three cloud backends:
- The Kconfig help text has been updated with further details on the contents of the certificate header.
- There are now default certificates headers in each backend that include the necessary variables with corresponding comments that document a typical PEM format certificate structure.
- Samples showcasing run-time provisioning (AWS, Azure and ATv2) has been updated to align with the aforementioned changes.

File: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (simensrostad) [reply subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:21]: I decided to keep this. If customers want to import their own certificates with corresponding naming they need this feature.
If not, they are bound to the header names used in `<service>-certs.h`

File: applications/asset_tracker_v2/src/cloud/CMakeLists.txt (hunk 13 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/CMakeLists.txt]: Turns out, this change is not necessary with the recent changes. But, I needed to use
`zephyr_include_directories_ifdef` and `zephyr_include_directories` in the `CMakeLists.txt` files of the various sample/apps to properly import each `x-cert.pem` file. If I used the `target_` flavour the files were never found by the cloud backend. Is it ok this way? Or do you suggest an alternative path?

**Comment by simensrostad** [comment]
@divipillai This PR has a small changelog update.

**Review by divipillai — APPROVED** [review]

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 14]
```diff
@@ -329,6 +331,14 @@ static void message_handler(struct debug_msg_data *msg)
 			LOG_ERR("Failed starting module, error: %d", err);
 			SEND_ERROR(debug, DEBUG_EVT_ERROR, err);
 		}
+
+		/* Notify the rest of the application that it is connected to network
+		 * when building for QEMU x86.
+		 */
+		if (IS_ENABLED(CONFIG_BOARD_QEMU_X86)) {
+			{ SEND_EVENT(debug, DEBUG_EVT_QEMU_X86_INITIALIZED); }
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:339]: Yes. I get a redefinition error if it is not included.

```
In file included from /home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/debug_module.c:21:
/home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/debug_module.c: In function 'message_handler':
/home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/modules_common.h:44:32: error: redefinition of 'event'
   44 |  struct _mod ## _module_event *event = new_ ## _mod ## _module_event();  \
      |                                ^~~~~
/home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/debug_module.c:340:4: note: in expansion of macro 'SEND_EVENT'
  340 |    SEND_EVENT(debug, DEBUG_EVT_QEMU_X86_NETWORK_CONNECTED);
      |    ^~~~~~~~~~
/home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/modules_common.h:44:32: note: previous definition of 'event' was here
   44 |  struct _mod ## _module_event *event = new_ ## _mod ## _module_event();  \
      |                                ^~~~~
/home/simensr/ncs/nrf/applications/asset_tracker_v2/src/modules/debug_module.c:339:4: note: in expansion of macro 'SEND_EVENT'
  339 |    SEND_EVENT(debug, DEBUG_EVT_QEMU_X86_INITIALIZED);
  ```
